In [1]:
import pandas as pd
import matplotlib

# Data Exploration

### __Read data to create dataframes__

In [2]:
brands = pd.read_csv('brands.csv')
orderlines = pd.read_csv('orderlines.csv')
orders = pd.read_csv('orders.csv')
products = pd.read_csv('products.csv')

### __df.info__

In [3]:
print(brands.info())
print(orderlines.info())
print(orders.info())
print(products.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   short   187 non-null    object
 1   long    187 non-null    object
dtypes: object(2)
memory usage: 3.0+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Col

### __Inspect dataframes heads and tails__

In [4]:
print('###################### BRANDS \n', brands.head(),'\n', brands.tail(),'\n---')
print('###################### ORDERLINES \n', orderlines.head().to_string(),'\n', orderlines.tail().to_string(),'\n---')
print('###################### ORDERS \n', orders.head().to_string(), '\n', orders.tail().to_string(),'\n---')
print('###################### PRODUCTS \n', products.head().to_string(), '\n', products.tail().to_string(),'\n---')

###################### BRANDS 
   short       long
0   8MO  8Mobility
1   ACM       Acme
2   ADN     Adonit
3   AII      Aiino
4   AKI     Akitio 
     short      long
182   XOO    Xoopar
183   XRI    X-Rite
184   XTO     Xtorm
185   ZAG  ZaggKeys
186   ZEP      Zepp 
---
###################### ORDERLINES 
         id  id_order  product_id  product_quantity      sku unit_price                 date
0  1119109    299539           0                 1  OTT0133      18.99  2017-01-01 00:07:19
1  1119110    299540           0                 1  LGE0043     399.00  2017-01-01 00:19:45
2  1119111    299541           0                 1  PAR0071     474.05  2017-01-01 00:20:57
3  1119112    299542           0                 1  WDT0315      68.39  2017-01-01 00:51:40
4  1119113    299543           0                 1  JBL0104      23.74  2017-01-01 01:06:38 
              id  id_order  product_id  product_quantity      sku unit_price                 date
293978  1650199    527398           0   

### __Gather some basic descriptives for all the dataframes__

#### brands

In [5]:
print('\n################################## .describe() \n\n', brands.describe())
print('\n################################## .isna().any() \n\n', brands.isna().any())
print('\n################################## .shape \n\n', brands.shape)
print('\n################################## .columns \n\n', brands.columns)


################################## .describe() 

        short    long
count    187     187
unique   187     181
top      8MO  Mophie
freq       1       2

################################## .isna().any() 

 short    False
long     False
dtype: bool

################################## .shape 

 (187, 2)

################################## .columns 

 Index(['short', 'long'], dtype='object')


#### orderlines



In [77]:
print(orderlines.head().to_string())
print('\n################################## .describe() \n\n', orderlines.describe())
print('\n################################## .isna().any() \n\n', orderlines.isna().any())
print('\n################################## .shape \n\n', orderlines.shape)
print('\n################################## .columns \n\n', orderlines.columns)

        id  order_id  product_id  product_quantity      sku  unit_price                date  total_price
0  1119109    299539           0                 1  OTT0133       18.99 2017-01-01 00:07:19        18.99
1  1119110    299540           0                 1  LGE0043      399.00 2017-01-01 00:19:45       399.00
2  1119111    299541           0                 1  PAR0071      474.05 2017-01-01 00:20:57       474.05
3  1119112    299542           0                 1  WDT0315       68.39 2017-01-01 00:51:40        68.39
4  1119113    299543           0                 1  JBL0104       23.74 2017-01-01 01:06:38        23.74

################################## .describe() 

                  id       order_id  product_id  product_quantity  \
count  2.939830e+05  293983.000000    293983.0     293983.000000   
mean   1.397918e+06  419999.116544         0.0          1.121126   
std    1.530096e+05   66344.486479         0.0          3.396569   
min    1.119109e+06  241319.000000         0.0 

#### orders

In [7]:
print(orders.head())
print('\n################################## .describe() \n\n', orders.describe())
print('\n################################## .isna().any() \n\n', orders.isna().any())
print('\n################################## .shape \n\n', orders.shape)
print('\n################################## .columns \n\n', orders.columns)
print('\n################################## .state.unique() \n\n', orders.state.unique())
print('\n################################## .order_id.duplicated \n\n', orders.order_id.duplicated().sum())

   order_id         created_date  total_paid      state
0    241319  2017-01-02 13:35:40       44.99  Cancelled
1    241423  2017-11-06 13:10:02      136.15  Completed
2    242832  2017-12-31 17:40:03       15.76  Completed
3    243330  2017-02-16 10:59:38       84.98  Completed
4    243784  2017-11-24 13:35:19      157.86  Cancelled

################################## .describe() 

             order_id     total_paid
count  226909.000000  226904.000000
mean   413296.482480     569.225818
std     65919.250331    1761.778002
min    241319.000000       0.000000
25%    356263.000000      34.190000
50%    413040.000000     112.990000
75%    470553.000000     525.980000
max    527401.000000  214747.530000

################################## .isna().any() 

 order_id        False
created_date    False
total_paid       True
state           False
dtype: bool

################################## .shape 

 (226909, 4)

################################## .columns 

 Index(['order_id', 'created_da

#### products

In [34]:
print(products.head().to_string())
print('\n################################## .describe() \n\n', products.describe())
print('\n################################## .isna().any() \n\n', products.isna().any())
print('\n################################## .shape \n\n', products.shape)
print('\n################################## .columns \n\n', products.columns)

       sku                                           name                                                           desc  price promo_price  in_stock      type
0  RAI0007              Silver Rain Design mStand Support                   Aluminum support compatible with all MacBook  59.99     499.899         1      8696
1  APP0023              Apple Mac Keyboard Keypad Spanish                      USB ultrathin keyboard Apple Mac Spanish.     59     589.996         0  13855401
2  APP0025               Mighty Mouse Apple Mouse for Mac                                         mouse Apple USB cable.     59     569.898         0      1387
3  APP0072  Apple Dock to USB Cable iPhone and iPod white                          IPhone dock and USB Cable Apple iPod.     25     229.997         0      1230
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM  2GB RAM Mac mini and iMac (2006/07) MacBook Pro (2006/07/08).  34.99       31.99         1      1364

################################## .des

# __Datacleaning__

## Cleaning orders

In [3]:
# Convert datatypes
orders['created_date'] = pd.to_datetime(orders['created_date'])

# Create new columns for 'year' (int), 'month' (int), 'day_of_month' (int), 'weekday' (categorical variable, ordered)
orders['year'] = pd.DatetimeIndex(orders['created_date']).year
orders['month'] = pd.DatetimeIndex(orders['created_date']).month
orders['day_of_month'] = pd.DatetimeIndex(orders['created_date']).day
orders['weekday'] = orders['created_date'].dt.strftime('%A')

# create sorted categories for weekday
cats = [ 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
orders['weekday'] = pd.Categorical(orders['weekday'], categories = cats, ordered=True)

# Exclude missing values in total_paid column (cause for missing values is order-state = pending) and check for na before and after manipulation
print(f'Nan-count before removal = {orders.total_paid.isna().sum()}')
orders.dropna(inplace=True)
print(f'Nan-count after removal = {orders.total_paid.isna().sum()}')

Nan-count before removal = 5
Nan-count after removal = 0


## Cleaning orderlines

In [4]:
# Convert datatypes
orderlines['date'] = pd.to_datetime(orderlines['date'])

# rename id_order to order_id to standardize it
orderlines.rename(columns={'id_order':'order_id'}, 
                   inplace=True)

In [5]:
# FIX UNIT_PRICE COLUMN DECIMAL ISSUE
## remove all decimal dots 
orderlines = orderlines.assign(unit_price_nd = orderlines['unit_price'].str.replace('\.','', regex=True))

## separate all numbers before and after the decimal dot
orderlines['integers'] = orderlines['unit_price_nd'].str[:-2]
orderlines['decimals'] = orderlines['unit_price_nd'].str[-2:]

## concatenate integers and decimals with dot in between
orderlines['new_unit_price'] = orderlines['integers'] + '.' + orderlines['decimals']

## create converted unit_price column based on fixed column
orderlines['unit_price'] = pd.to_numeric(orderlines['new_unit_price'])

## drop 'auxiliary' columns
orderlines.drop(['unit_price_nd','decimals','integers','new_unit_price'], axis=1, inplace=True)

# create a new column "total_price" by multiplying product_quantity times unit_price
orderlines['total_price'] = orderlines['product_quantity'] * orderlines['unit_price']

## Cleaning products

In [6]:
# Duplicates
print(f'Duplicates in products: {products.duplicated("sku").sum()}')
# --> 8747
products.drop_duplicates(subset= 'sku', inplace=True)
print(f'Duplicates in products after removal: {products.duplicated("sku").sum()}')

Duplicates in products: 8747
Duplicates in products after removal: 0


In [7]:
# Replace all price values in products that have: 
# --> more than three digits after the dot
# --> missing value
# with the unit_price value from orderlines
# returns a df with two columns, 'sku' and 'price'

price_from_orderlines = (
    products
        .merge(
                orderlines
                        .groupby('sku')
                        .agg(max_price = ('unit_price', 'max')), 
                how = 'left', 
                on= 'sku'
                )
        .assign(price = lambda df: df['max_price'].round(2) if (lambda df_ : (df_['price'].str.contains("('.*\.\d{3}$')", regex=True)) | (df_['price'].isna()))  else df['price'])
        .dropna(subset = 'price')
)

In [8]:
# # Fixing price column

# Fixing decimal point issue
# create a new column with amount of dots in the price column
products['dots']=products['price'].str.count('.')

# remove all decimal dots 
products = (
    products
        .assign(price_nd = lambda df: df['price'].str.replace('\.','', regex=True) if type(df['price'] == 'string') else df['price'])
        )

# separate all numbers before and after the decimal dot
products['integers'] = products['price_nd'].str[:-2]
products['decimals'] = products['price_nd'].str[-2:]

# if we dont have dots makes the decimals 00
products.loc[products['dots']==0,'decimals']='00'
products.loc[products['dots']==0,'integers']=products['price_nd']

# concatenate integers and decimals with dot in between
products['new_price'] = products['integers'] + '.' + products['decimals']

# create converted price column based on fixed column
products['price'] = pd.to_numeric(products['new_price'])

# drop 'auxiliary' columns
products.drop(['price_nd','decimals','integers','new_price', 'dots'], axis=1, inplace=True)

# products.head(15)

In [18]:
# Insert price values from orderlines back into products table and remove missing values

dict_price_orderlines = dict(price_from_orderlines[['sku', 'price']].values)
products = (
    products
        .assign(price = lambda df: df['sku'].map(dict_price_orderlines) if (lambda df_: df_['sku'] in dict_price_orderlines.keys()) else df['price'])
        .dropna(subset = 'price')
    )


sku                                               name  \
0      RAI0007                  Silver Rain Design mStand Support   
1      APP0023                  Apple Mac Keyboard Keypad Spanish   
2      APP0025                   Mighty Mouse Apple Mouse for Mac   
3      APP0072      Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                                ...   
19279  MMW0012  My MW Case MacBook Pro 13 "(Late 2016) Blue Sa...   
19280  MMW0013  My MW Case MacBook Pro 13 "(Late 2016) White S...   
19281  MMW0016       My MW Case MacBook Pro 13 "(Late 2016) Black   
19282  MMW0015        My MW Case MacBook Pro 13 "(Late 2016) Gray   
19320  REP0076           repair Full screen iPad (1st generation)   

                                                    desc   price promo_price  \
0           Aluminum support compatible with all MacBook   54.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.   60.49     589.996   
2                                 mouse Apple USB cable.   59.00     569.898   
3                  IPhone dock and USB Cable Apple iPod.   25.00     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...   35.71       31.99   
...                                                  ...     ...         ...   
19279  Avoid shock and damage to your MacBook Pro 13-...   22.99       20.99   
19280  Avoid shock and damage to your MacBook Pro 13-...   22.99       20.99   
19281  Avoid shock and damage to your MacBook Pro 13-...   22.99       20.99   
19282  Avoid shock and damage to your MacBook Pro 13-...   22.99       20.99   
19320  Repair service including parts and labor for iPad  149.99   1.499.795   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19279         0  13835403  
19280         1  13835403  
19281         1  13835403  
19282         1  13835403  
19320         0  1,44E+11  

[7850 rows x 7 columns]

In [10]:
# Remove all dots from promo_prices and add one dot before the second to last digit, then convert to numeric
products = (
    products
        .assign(promo_price = lambda df: df['promo_price'].str.replace('.', ''))
        .assign(promo_price = lambda df: df['promo_price'].str[:-2] + '.' + df['promo_price'].str[-2:])
        .dropna(subset = 'promo_price')
        .assign(promo_price = lambda df: pd.to_numeric(df['promo_price']))
)

C:\Users\tobia\AppData\Local\Temp\ipykernel_21424\2699542818.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  .assign(promo_price = lambda df: df['promo_price'].str.replace('.', ''))


In [11]:
# Function to reduce promo_price
# def reduce_promo_price(row):
#     if row['promo_price'] < row['price']:
#         row['price'] = row['price'] / 10

(
products
    .assign(promo_price = lambda df: (df['promo_price'] / 10) if )
)


       sku                               name  \
0  RAI0007  Silver Rain Design mStand Support   
1  APP0023  Apple Mac Keyboard Keypad Spanish   
2  APP0025   Mighty Mouse Apple Mouse for Mac   

                                           desc  price  promo_price  in_stock  \
0  Aluminum support compatible with all MacBook  59.99      4998.99         1   
1     USB ultrathin keyboard Apple Mac Spanish.   0.59      5899.96         0   
2                        mouse Apple USB cable.   0.59      5698.98         0   

       type  
0      8696  
1  13855401  
2      1387  


AttributeError: Can only use .str accessor with string values!

In [63]:
# TESTING CELL

# (
# products
#     .dropna()
#     .query("price.str.len() > 6", engine = 'python')
#     .head(15)   
# )
# (
# products
#     .dropna()
#     .count()
# )

# (
# products
#     .dropna()
#     .query("price.str.match('.*\.\d{3}$')")
#     .head(25)
# )


AttributeError: Can only use .str accessor with string values!

### Price and Column Price Columns
* First idea is to remove all rows containing prices that have three digits after the last dot
* For now dropping all na
* Checking the number of rows that have this issue: 
    * price: 537 / 10477 total rows
    * promo_price: 9638 / 10477 total rows

## Cleaning brands

# Data-Consistency Checks

## Dates between orders and orderlines

In [12]:
# Check whether dates of first and last orders match in orders vs orderlines

min_date_orders=min(orders['created_date'])
max_date_orders= max(orders['created_date'])
min_date_orderlines = min(orderlines['date'])
max_date_orderlines = max(orderlines['date'])

pd.DataFrame(
    {
        'df': ['orders', 'orderlines'],
        'first_order': [min_date_orders, min_date_orderlines],
        'last_order': [max_date_orders, max_date_orderlines]
    })

df         first_order          last_order
0      orders 2017-01-01 00:07:19 2018-03-14 13:58:36
1  orderlines 2017-01-01 00:07:19 2018-03-14 13:58:36

## Total price between orders and orderlines

In [13]:
# Exploration of price mismatch from perspective of ORDERLINES

print(f'Total price sum in orders=     {orders.total_paid.sum()}\nTotal price sum in orderlines= {orderlines.total_price.sum()}\n')

# Exploration of price mismatch from perspective of ORDERLINES
## create copy of orderlines with column checking whether order id is also in orders df
ol_copy = orderlines.assign(check_orders = orderlines['order_id'].isin(orders['order_id'])).copy()

# check unequal rows
print(
ol_copy
    .query("check_orders==False")
)
print()

# there are 240 rows of orders that are not in the orders df - these will be removed for now
orderlines = (ol_copy
              .query("check_orders==True").copy())

Total price sum in orders=     129159615.07000001
Total price sum in orderlines= 128776222.02999999

             id  order_id  product_id  product_quantity      sku  unit_price  \
5       1119114    295310           0                10  WDT0249      231.79   
63      1119218    296284           0                 1  BNQ0042      699.00   
67      1119223    294806           0                 1  APP1849     2558.99   
69      1119226    294806           0                 1  APP1864     2797.99   
70      1119235    297261           0                 1  QNA0177      304.99   
...         ...       ...         ...               ...      ...         ...   
275665  1621177    244328           0                 1  OWC0260      349.00   
280055  1628699    261391           0                 1  APP2352     3343.00   
280352  1629247    287797           0                 1  SYN0182      484.11   
280856  1630150    261391           0                 1  SAT0091       44.99   
280869  1630178    

In [14]:
# Exploration of price mismatch from perspective of ORDERS
# check whether order id is also in orderlines df
(
orders
 .assign(check_orders = orders['order_id'].isin(orderlines['order_id']))
 .query("check_orders==False")
)

# check state of orders that are in orders but not in orderlines
(
orders
    .assign(check_orders = orders['order_id'].isin(orderlines['order_id']))
    .query("check_orders==False")
    ['state'].value_counts()
)

# Can be explored further, but for now all unmatched orders will be dropped
orders = (orders
            .assign(check_orders = orders['order_id'].isin(orderlines['order_id']))
            .query("check_orders==True")
         )

In [15]:
# Merge orderlines and order to check price-differences
ol_orders_merged = (
    orderlines
        .groupby('order_id')
        .agg({'total_price':'sum'})
        .merge(orders, how='left', on='order_id')
        .rename(columns= {
            'total_price':'total_price_orderlines',
            'total_paid': 'total_paid_orders'
        })
        .copy()
)

# create price-difference column and remove small differences
def get_price_difference(row):
    if (abs(row['total_price_orderlines'] - row['total_paid_orders']) < 0.01):
        result = 0
    else:
        result = (row['total_price_orderlines'] - row['total_paid_orders'])
    return result

ol_orders_merged['price_diff'] = ol_orders_merged.apply(lambda row : get_price_difference(row), axis=1)


In [16]:
# Explore price differences

# (
# ol_orders_merged
#     .query("price_diff != 0 & state == 'Completed'")
#     .price_diff.describe()
# )

(
ol_orders_merged
    .query("price_diff != 0 & state == 'Completed'")
    .sort_values(by='price_diff')
    .head(10)
)

order_id  total_price_orderlines        created_date  \
145       297148                   84.55 2017-01-01 16:42:24   
123       293308                   66.49 2017-01-01 13:33:43   
191738    512894                 1674.60 2018-02-16 13:25:30   
188176    508825                 1291.60 2018-02-16 11:21:44   
192231    513447                   94.98 2018-02-17 19:26:59   
191838    513009                   20.00 2018-02-16 17:33:09   
191134    512146                  313.00 2018-02-15 17:32:17   
188923    509673                 2408.99 2018-02-17 16:25:44   
191874    513049                  249.00 2018-02-16 19:03:22   
191804    512969                  123.99 2018-02-16 15:48:22   

        total_paid_orders      state  year  month  day_of_month   weekday  \
145               4069.54  Completed  2017      1             1    Sunday   
123               2635.47  Completed  2017      1             1    Sunday   
191738            3356.18  Completed  2018      2            16    Friday   
188176            2590.18  Completed  2018      2            16    Friday   
192231             486.89  Completed  2018      2            17  Saturday   
191838             406.80  Completed  2018      2            16    Friday   
191134             632.98  Completed  2018      2            15  Thursday   
188923            2708.97  Completed  2018      2            17  Saturday   
191874             504.98  Completed  2018      2            16    Friday   
191804             376.96  Completed  2018      2            16    Friday   

        check_orders  price_diff  
145             True    -3984.99  
123             True    -2568.98  
191738          True    -1681.58  
188176          True    -1298.58  
192231          True     -391.91  
191838          True     -386.80  
191134          True     -319.98  
188923          True     -299.98  
191874          True     -255.98  
191804          True     -252.97

### Insights on price-differences

* by far the largest price-differences exist for orders that were not completed - maybe people who put a huge amount of products into their shopping cart
* price differences that were less than +- 1 cent were set to 0 to clean the dataset 
* recommendation: remove all rows containing price-differences that are higher or lower than the highest or lowest order that was completed
  * High cap = > 165.00 
  * Low cap  = < -3984.99

### Removal of orders with largest price-differences 

In [17]:
ol_orders_merged = (
    ol_orders_merged
        .query("(price_diff >= -3984.99) & (price_diff <= 165)")
)

# Export of orders and orderlines

In [18]:
orderlines.to_csv(R'C:\Users\tobia\Documents\Bootcamp\03_datacleaning_storytelling_pd_matplot\00_eniac_data\cleaned_data\orderlines_cleaned.csv', index=False)
orders.to_csv(R'C:\Users\tobia\Documents\Bootcamp\03_datacleaning_storytelling_pd_matplot\00_eniac_data\cleaned_data\orders_cleaned.csv', index=False)

# __Exploration of specific questions__

## Suggested Questions

__How many orders are there?__  
226909 unique order_ids.  
  
__How many products are there?__  
10579 unique products  

__Which timespan is captured in the dataset?__  
First order in dataset made on: 2017-01-01 00:07:19  
Last order in dataset made on: 2018-03-14 13:58:36

__How many orders are completed?__  
46605 orders are completed.  
  
__How should revenue be computed?__  
Summarize price of all unique orders with state = completed.  
Total revenue from captured data: 15910079.24 €



In [19]:
# num of orders
orders.order_id.nunique()
# num of products
products.sku.nunique()

# timeperiod captured
first_order = orders.created_date.min()
last_order = orders.created_date.max()
print(f'First order in dataset made on: {first_order}\nLast order in dataset made on: {last_order}')

# num of completed orders
orders.loc[orders['state'] == 'Completed','order_id'].count()

# revenue
orders.loc[orders['state'] == 'Completed','total_paid'].sum()

First order in dataset made on: 2017-01-01 00:07:19
Last order in dataset made on: 2018-03-14 13:58:36


15907059.889999995

## Revenue  
  
__Monthly Revenue__:  


In [20]:
rev_by_day = pd.DataFrame(
    orders
    .loc[orders['state'] == 'Completed', 'total_paid']
    .groupby(orders.day_of_week)
    .sum()
)
rev_by_day.reset_index(inplace=True)

rev_by_day['day_of_week_cat'] = pd.Categorical(
    rev_by_day['day_of_week'],
    categories=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
    ordered = True
)
rev_by_day = rev_by_day.sort_values('day_of_week_cat')

print(rev_by_day.head(10))

rev_by_day.plot.bar(x = 'day_of_week_cat')


AttributeError: 'DataFrame' object has no attribute 'day_of_week'